torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -384,235 +401,164 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::allreduce(
   std::unique_lock<std::mutex> cudaFreeMutexLock(
       *(c10::cuda::CUDACachingAllocator::getFreeMutex()));
 
+  pre(ncclStreams_[key]);
+
   C10D_NCCL_CHECK(ncclGroupStart());
 
-  for (size_t i = 0; i < tensors.size(); ++i) {
+  for (size_t i = 0; i < inputs.size(); ++i) {
     gpuGuard.set_index(devices[i].index());
     at::cuda::CUDAStream& ncclStream = ncclStreams_[key][i];
 
-    // Input `tensors` are created on a worker stream and used in different
-    // ncclStream. Hence, `tensors` must record the ncclStream to prevent being
-    // freed before ncclAllReduce finishes. See [Sync Streams].
+    // Both `inputs' and `outputs' are created on a worker stream and used in
+    // different ncclStreams.  Hence, both must record the ncclStream to
+    // prevent being freed before the collective finishes.
+    //
+    // We only record `inputs' here, and leave recording `outputs' to `fn' for
+    // operations where `inputs' and `outputs' are the same.
```

Comment:
"[...] where inputs and outputs are **NOT** the same.", otherwise it would be redundant, no?

Reply:
Yep—will fix.